Docs/post for release 23.2 (#8)
modelix/website: content/en/blog/2023-11-modelix-platform-release-23.2/index.md

## Model Client as NPM Package (experimental)

The previously existing [model client](https://github.com/modelix/modelix.core/tree/modelix-23.2/model-client) is now also published as [`@modelix/model-client` NPM package](https://artifacts.itemis.cloud/service/rest/repository/browse/npm-open/%40modelix/model-client/).

The exposed API is minimal and experimental. 
In the future, we want to make more features from the Kotlin model client directly accessible from the exposed API in the NPM package.

## Vue.js Bindings (experimental)

Built upon [`@modelix/model-client`](https://artifacts.itemis.cloud/service/rest/repository/browse/npm-open/%40modelix/model-client/) we created a library to interact with models through [Vue.js](https://vuejs.org/).

With [`@modelix/vue-model-api`](https://docs.modelix.org/modelix/23.2/core/reference/component-vue-model-api.html) we are currently building sample projects with Vue.js.
After that, we want to stabilize the API and show you how you can use it for your projects.

## MPS Model Adapters

> We created [adapters](https://github.com/modelix/modelix.core/tree/modelix-23.2/mps-model-adapters) that serve as an interface between MPS and the modelix.

Reviewer: ```suggestion
We created [adapters](https://github.com/modelix/modelix.core/tree/modelix-23.2/mps-model-adapters) that serve as an interface between MPS and the modelix model-api.
```

Author: I simplified the sentence further.

"We created [adapters](https://github.com/modelix/modelix.core/tree/modelix-23.2/mps-model-adapters) between MPS and modelix."

> They forward changes made to a model through the [model-api of modelix](https://docs.modelix.org/modelix/23.2/core/reference/component-model-api.html) to the [Open API of MPS](https://www.jetbrains.com/help/mps/open-api-accessing-models-from-code.html) of MPS and vice versa.

Reviewer: ```suggestion
They forward changes made to a model through the [model-api of modelix](https://docs.modelix.org/modelix/23.2/core/reference/component-model-api.html) to the [Open API of MPS](https://www.jetbrains.com/help/mps/open-api-accessing-models-from-code.html) of MPS.
```
The other direction is not available (yet) and might be limited compared to the primary one.

Author: Done

---
date: 2023-11-02
title: "modelix platform release 23.2"
---

> We are happy to announce the Autumn release of the modelix platform: **`modelix-23.2`**

Reviewer 2: ```suggestion
We are happy to announce the autumn release of the modelix platform: **`modelix-23.2`**
```

Author: Done

> The version bundled in this release are:

Reviewer: ```suggestion
The versions bundled in this release are:
```

Author: Done

```
modelixCoreVersion = 3.12.1
incrementalVersion = 0.2.0
buildToolsVersion = 1.1.0
modelixMPSPatchVersion = 157
```

> If you want to know how to use this platform in your Gradle build, check out the corresponding [Documentation](https://docs.modelix.org/modelix/main/platform/howto/usage-platform.html).

Reviewer 3: this should point to the released docs instead of main

Author: Done

- [Feature Highlights](#feature-highlights)
  - [ModelQL](#modelql)
  - [Model client as NPM package (alpha)](#model-client-as-npm-package-alpha)
  - [Vue.js Bindings (alpha)](#vuejs-bindings-alpha)
  - [MPS Model Adapters](#mps-model-adapters)
  - [Bulk Model Synchronization Plugin (alpha)](#bulk-model-synchronization-plugin-alpha)
- [Development Insights](#development-insights)
  - [New approach to MPS related components](#new-approach-to-mps-related-components)
  - [Contribution guide](#contribution-guide)
  - [Marking experimental features](#marking-experimental-features)
  - [Work on code quality](#work-on-code-quality)
- [What's to come](#whats-to-come)

# Feature Highlights

## ModelQL

> The [ModelQL query language](https://docs.modelix.org/modelix/modelix-23.2/core/explanation/modelql.html) provides a dynamic way of loading parts of the model on demand while reducing the number of requests to a minimum.

Reviewer 3: Leave a comment that this is what often is referred to as a 'delta API'

Reviewer 2: Really? Doesn't ModelQl always fetch all data needed, independent of how the data emerged out of edit deltas?

Reviewer 3: This is true. I guess I wanted to somehow cover the missing rationale for a 'why ModelQL?' in this. From my perspective we can leave it as is.

> The [ModelQL query language](https://docs.modelix.org/modelix/modelix-23.2/core/explanation/modelql.html) provides a dynamic way of loading parts of the model on demand while reducing the number of requests to a minimum.

Author: Replace `modelix-23.2` to `23.2` in links.

Author: Done.

> ## Model client as NPM package (alpha)

Reviewer 2: ```suggestion
## Model client as NPM package (experimental/unstable)
```

To be consistent with the unstable annotation and news?

Author: Done (decided for "experimental")

The previously existing [model client](https://github.com/modelix/modelix.core/tree/modelix-23.2/model-client) is now also published as [`@modelix/model-client` NPM package](https://artifacts.itemis.cloud/service/rest/repository/browse/npm-open/%40modelix/model-client/).

> The exposed API is minimal and experimental. In the future, we want to make more features from the Kotlin model client library directly accessible. 

Reviewer 3: Make it accessible where? It might be good to be clear here where / who will consume this API

Author: Done

> ## Vue.js Bindings (alpha)

Reviewer 2: Casing is inconsistent. "Model client as NPM package" uses sentence case (which I prefer due to being better readable for most readers). Other headings use title case.

Author: Done.
Decided for title case because the other blog posts also use it for their headings.

> Built upon [`@modelix/model-client`](https://artifacts.itemis.cloud/service/rest/repository/browse/npm-open/%40modelix/model-client/) we created a library to interact with models through Vue.js.

Reviewer 2: Link Vue.js once here?

Author: Done

> With [`@modelix/vue-model-api`](https://docs.modelix.org/modelix/modelix-23.2/core/reference/component-vue-model-api.html) we are currently building sample projects with Vue.js. After that, we want to stabilize the API and show you how you can use it for your projects.

Reviewer 2: ```suggestion
With [`@modelix/vue-model-api`](https://docs.modelix.org/modelix/modelix-23.2/core/reference/component-vue-model-api.html) we are currently building sample projects with Vue.js.
After that, we want to stabilize the API and show you how you can use it for your projects.
```

https://sive.rs/1s

Author: Done.
(Nice blog btw., will read more posts)

## MPS Model Adapters
We created [adapters](https://github.com/modelix/modelix.core/tree/modelix-23.2/mps-model-adapters) that serve as an interface between MPS and the modelix model-api.

> They are compatible with all MPS versions that modelix currently supports.

Reviewer 2: Is it really worth mentioning this in the blog post? This is for most users only an internal detail of modelix.

Author: I think it might be interesting for some users, so I would rather leave it in.

Reviewer 2: Those not knowing what that is are left a bit puzzled. Maybe we can then add one sentence explaining for which purpose these adapters are useful?

Author: Done. I hope the added description makes it more understandable.

@mhuster23 Please check if my description is accurate.

Reviewer: Left some comments

## Bulk Model Synchronization Plugin (alpha)

> The [bulk-model-sync Gradle Plugin](https://docs.modelix.org/modelix/modelix-23.2/core/reference/component-bulk-model-sync-gradle.html) allows bulk synchronization of models between MPS and a model-server.

Reviewer 2: Link to model-server documentation?

Author: Done

> Changes are applied incrementally and it can be run in existing CI systems to schedule and run the synchronization in both directions.

Reviewer 2: "it" is a bit unclear. Better spell that out.

Author: Done

This way you can implement a custom Git Integration tailored to your needs.

> The plugin is currently considered unstable, but will be stabilized in the coming weeks.

Reviewer 2: ```suggestion
The plugin is currently considered unstable, but will be stabilized soon.
```

Maybe this give a bit more freedom? ;-)

Author: Done

# Development Insights

> ## New approach to MPS related components

Reviewer 2: ```suggestion
## New approach to MPS-related components
```

Author: Done

> Up until now, we built and released MPS specific components for each MPS version individually.

Reviewer 2: ```suggestion
Up until now, we built and released MPS-specific components for each MPS version individually.
```

Author: Done

> Going forward, we want to release single components, which are compatible with all MPS versions, that are currently supported by modelix.

Reviewer 2: The "why" is not answered here. Something like "to reduce maintenance efforts" should be added here.

Author: Done.

The first component to follow this new approach is [mps-model-adapters](#mps-model-adapters)

To achieve that, we separate code in MPS solutions from support library code wherever possible.

> Also, we are developing upcoming MPS plugins as IntelliJ plugins.

Reviewer 3: ```suggestion
Also, we are developing upcoming MPS plugins as IntelliJ plugins instead of MPS plugins, reducing the previously needed overhead when updating and maintaining such plugins.
```

Author: > MPS plugins as IntelliJ plugins instead of MPS plugins

@nkoester Would it be correct to say the following?

"Also, we are developing upcoming MPS plugins as IntelliJ plugins instead of MPS **solutions**, reducing the previously needed overhead when updating and maintaining such plugins."

Reviewer 3: Not quite, I would stick to MPS plugins. Yes, MPS plugins are 'solutions', but there is a lot of things hacked around it internally. MPS plugins means much more in this context.

Author: Done, but split in two sentences for readability.

## Contribution guide

> We created a [contribution guide](https://docs.modelix.org/modelix/main/main/contribute.html) to guide onboarding and external contributors.

Reviewer 2: a bit too much "guide" in this sentence ;-)

```suggestion
We created a [contribution guide](https://docs.modelix.org/modelix/main/main/contribute.html) to help new contributors.
```

Author: Done

Author: Done

## Marking experimental features

> We decided to mark features sometimes as experimental so that we could publish them but still iterate quickly with ideas and APIs.

Reviewer 2: ```suggestion
From now on, we mark experimental features clearly.
This allows the modelix team to publish them soon for testing and iterating quickly with ideas and APIs and users can try those out while being aware of the experimental nature.
```

Author: Done

> A separate blog post will be published detailing the annotations used in Kotlin and TypeScript that mark features as experimental.

Reviewer 2: Why do we actually need this blog post? As a reader I'd directly want to read this now. We can easily link this section to anyone interested. Or both this entry and the other one are published in parallel to make this section linkable.

Author: Writing a separate blog post is overkill.
I ditched the separate post https://github.com/modelix/website/pull/9 and linked the documentation here.

## Work on code quality

> In our automation, we introduced static code analysis, dependency updating and security checking.

Reviewer 2: Should also answer the "why" question.

Author: Done (also changed subheading accordingly)

# What's to come

> Unfortunately, we were not able to update the roadmap before releasing modelix-23.2. We are dedicated to bring an updated version during December 2023 after our already scheduled planning sessions. Please check the [latest roadmap in our documentation](https://docs.modelix.org/modelix/main/main/roadmap.html).

Reviewer 2: ```suggestion
Unfortunately, we were not able to update the roadmap before releasing modelix-23.2.
We are dedicated to bring an updated version during December 2023 after our already scheduled planning sessions.
Please check the [latest roadmap in our documentation](https://docs.modelix.org/modelix/main/main/roadmap.html).
```

Do we want to publish that many details about our way of working?

Author: Done applying new lines.

Author: > Do we want to publish that many details about our way of working?

@nkoester @mhuster23 I think this was the consensus in our weekly?
We want to clearly communicate, why we did not update the road map.

Reviewer 3: While I want to share that we will provide a roadmap this year, I can understand that we might want to reduce the verbosity here. Maybe just remove the details and keep the overall message that there will be a roadmap update in December.

Author: Done

## MPS Model Adapters

We created [adapters](https://github.com/modelix/modelix.core/tree/23.2/mps-model-adapters) that serve as an interface between MPS and the modelix [model-api](https://docs.modelix.org/modelix/23.2/core/reference/component-model-api.html).
They are compatible with all MPS versions that modelix currently supports.

## Bulk Model Synchronization Plugin (experimental)

The [bulk-model-sync Gradle Plugin](https://docs.modelix.org/modelix/23.2/core/reference/component-bulk-model-sync-gradle.html) allows bulk synchronization of models between MPS and a [model-server](https://docs.modelix.org/modelix/23.2/core/reference/component-model-server.html).
The plugin applies changes incrementally, and can be run in existing CI systems to schedule and run the synchronization in both directions.
This way you can implement a custom Git Integration tailored to your needs.

The plugin is considered unstable, but will be stabilized soon.

# Development Insights

## New Approach to MPS-Related Components

Up until now, we built and released MPS-specific components for each MPS version individually.

> Going forward, we want to release single components, which are compatible with all MPS versions, that are currently supported by modelix.

Reviewer 2: ```suggestion
Going forward, we want to release single components, which are compatible with all MPS versions that are currently supported by modelix.
```

https://www.merriam-webster.com/grammar/when-to-use-that-and-which I have to read that up from time to time again because the rules are so unintuitive for Germans

Author: Done
